# Task 4.10 Part A Coding Etiquette and Excel Reporting

Contents

1. Import libraries


2. Import latest data set


3. Address PII data issues


4. Create region column and compare customer spending habits per region


5. Exclude low activity customers


6. Export adjusted dataframe as pickle


# 1. Import libraries

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 2. Import latest dataset

In [2]:
# Turn project folder path into a string

path = r'/Users/giorjeanmutuc/Instacart Basket Analysis'

In [3]:
# import ords_prods_cust_merge.pkl

ords_prods_cust_merge = pd.read_pickle(r'/Users/giorjeanmutuc/Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_cust_merge.pkl')

In [4]:
# remove column limit to enable checking the data

pd.options.display.max_columns = None

In [5]:
# Check column heads

ords_prods_cust_merge.head(5)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

       activity busiest_period_of_day  max_order  loyalty_flag  avg_order  \
0  Regular days        Average orders         10  New customer   6.367797   
1  Slowest days        Average orders         10  New customer   6.367797   
2  Slowest days           Most orders         10  New customer   6.367797   
3  Slowest days        Average orders         10  New customer   6.367797   
4  Slowest days           Most orders         10  New customer   6.367797   

  spending_flag  median_freq  ord_med_freq        order_freq  \
0   Low spender         20.0          20.0  Regular customer   
1   Low spender         20.0          20.0  Regular customer   
2   Low spender         20.0          20.0  Regular customer   
3   Low spender         20.0          20.0  Regular customer   
4   Low spender         20.0          20.0  Regular customer   

    order_freq_flag first_name surname  gender    state  age date_joined  \
0  Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
1  Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
2  Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
3  Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   
4  Regular customer      Linda  Nguyen  Female  Alabama   31   2/17/2019   

   dependents fam_status  income _merge  
0           3    married   40423   both  
1           3    married   40423   both  
2           3    married   40423   both  
3           3    married   40423   both  
4           3    married   40423   both

In [6]:
# Check number of rows and columns

ords_prods_cust_merge.shape

(32404859, 34)

In [7]:
# Check all kinds of variables

ords_prods_cust_merge.dtypes

order_id                     int64
user_id                      int64
order_number                 int64
order_dow                    int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
price_range_loc             object
activity                    object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
avg_order                  float64
spending_flag               object
median_freq                float64
ord_med_freq               float64
order_freq                  object
order_freq_flag             object
first_name                  object
surname                     object
gender                      object
state                       object
age                 

In [ ]:
# Change some columns to string to optimize .describe function

ords_prods_cust_merge[['order_id','user_id','product_id','aisle_id','department_id']] = ords_prods_cust_merge[['order_id','user_id', 'product_id', 'aisle_id', 'department_id']].astype(str)

In [9]:
# Check if columns have been converted to object

ords_prods_cust_merge[['order_id','user_id','product_id','aisle_id','department_id']].dtypes

order_id         object
user_id          object
product_id       object
aisle_id         object
department_id    object
dtype: object

In [10]:
# Finding missing values

ords_prods_cust_merge.isnull().sum()

order_id                        0
user_id                         0
order_number                    0
order_dow                       0
order_hour_of_day               0
days_since_prior_order          0
product_id                      0
add_to_cart_order               0
reordered                       0
product_name                    0
aisle_id                        0
department_id                   0
prices                       5127
price_range_loc                 0
activity                        0
busiest_period_of_day           0
max_order                       0
loyalty_flag                    0
avg_order                       0
spending_flag                   0
median_freq                     0
ord_med_freq                    0
order_freq                      0
order_freq_flag                 0
first_name                1775118
surname                         0
gender                          0
state                           0
age                             0
date_joined   

# Missing values

1775118 NAN first names. We can leave it as is since we're removing the first_column later. 
5127 NAN prices which are the outliers 14900 and 99999 renamed earlier.

In [11]:
# Format scientific notation

pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [12]:
# Check if outliers under Prices column has been changed to NAN

ords_prods_cust_merge['prices'].max()

25.0

# 3. Address PII data issues

In [4]:
# Delete first name and surname columns as these information infringe on privacy
# Delete repetitive and irrelevant columns

df_merge_PII = ords_prods_cust_merge.drop(columns = ['first_name', 'surname', 'order_freq', 'median_freq'])

In [14]:
# Check column headings

df_merge_PII.head(5)

order_id user_id  order_number  order_dow  order_hour_of_day  \
0  2539329       1             1          2                  8   
1  2398795       1             2          3                  7   
2   473747       1             3          3                 12   
3  2254736       1             4          4                  7   
4   431534       1             5          4                 15   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                   0.000        196                  1          0   
1                  15.000        196                  1          1   
2                  21.000        196                  1          1   
3                  29.000        196                  1          1   
4                  28.000        196                  1          1   

  product_name aisle_id department_id  prices    price_range_loc  \
0         Soda       77             7   9.000  Mid-range product   
1         Soda       77             7   9.000  Mid-range product   
2         Soda       77             7   9.000  Mid-range product   
3         Soda       77             7   9.000  Mid-range product   
4         Soda       77             7   9.000  Mid-range product   

       activity busiest_period_of_day  max_order  loyalty_flag  avg_order  \
0  Regular days        Average orders         10  New customer      6.368   
1  Slowest days        Average orders         10  New customer      6.368   
2  Slowest days           Most orders         10  New customer      6.368   
3  Slowest days        Average orders         10  New customer      6.368   
4  Slowest days           Most orders         10  New customer      6.368   

  spending_flag  ord_med_freq   order_freq_flag  gender    state  age  \
0   Low spender        20.000  Regular customer  Female  Alabama   31   
1   Low spender        20.000  Regular customer  Female  Alabama   31   
2   Low spender        20.000  Regular customer  Female  Alabama   31   
3   Low spender        20.000  Regular customer  Female  Alabama   31   
4   Low spender        20.000  Regular customer  Female  Alabama   31   

  date_joined  dependents fam_status  income _merge  
0   2/17/2019           3    married   40423   both  
1   2/17/2019           3    married   40423   both  
2   2/17/2019           3    married   40423   both  
3   2/17/2019           3    married   40423   both  
4   2/17/2019           3    married   40423   both

In [15]:
# Check number of rows and columns

df_merge_PII.shape

(32404859, 30)

# 4. Create region column and compare customer spending habits per region.


In [5]:
# Check current columns in the latest dataset

df_merge_PII.columns

Index(['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order', 'product_id', 'add_to_cart_order',
       'reordered', 'product_name', 'aisle_id', 'department_id', 'prices',
       'price_range_loc', 'activity', 'busiest_period_of_day', 'max_order',
       'loyalty_flag', 'avg_order', 'spending_flag', 'ord_med_freq',
       'order_freq_flag', 'gender', 'state', 'age', 'date_joined',
       'dependents', 'fam_status', 'income', '_merge'],
      dtype='object')

In [6]:
# create region column
# create a flag for Northeast region

df_merge_PII.loc[df_merge_PII['state'].isin(['Maine', 'New Hampshire', 'Vermont','Massachusetts','Rhode Island','Connecticut','New York','Pennsylvania','New Jersey']), 'region'] = 'Northeast' 

In [7]:
# create flag for Midwest region

df_merge_PII.loc[df_merge_PII['state'].isin(['Wisconsin','Michigan','Illinois','Indiana','Ohio','North Dakota','South Dakota','Nebraska','Kansas','Minnesota','Iowa','Missouri']), 'region'] = 'Midwest'

In [8]:
# create flag for South region

df_merge_PII.loc[df_merge_PII['state'].isin(['Delaware','Maryland','District of Columbia','Virginia','West Virginia','North Carolina','South Carolina','Georgia','Florida','Kentucky','Tennessee','Mississippi','Alabama','Oklahoma','Texas','Arkansas','Louisiana']), 'region'] = 'South' 

In [9]:
# create flag for West region

df_merge_PII.loc[df_merge_PII['state'].isin(['Idaho','Montana','Wyoming','Nevada','Utah','Colorado','Arizona','New Mexico','Alaska','Washington','Oregon','California','Hawaii']), 'region'] = 'West' 

In [21]:
df_merge_PII['region'].value_counts(dropna = False)

South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
Name: region, dtype: int64

In [22]:
# head results just for user_id, state & region

df_merge_PII[['user_id', 'state', 'region']].head(10)

user_id    state region
0       1  Alabama  South
1       1  Alabama  South
2       1  Alabama  South
3       1  Alabama  South
4       1  Alabama  South
5       1  Alabama  South
6       1  Alabama  South
7       1  Alabama  South
8       1  Alabama  South
9       1  Alabama  South

In [10]:
# Crosstab of region & spending_flag to see customer spending habits

crosstab = pd.crosstab(df_merge_PII['region'], df_merge_PII['spending_flag'], dropna = False)

In [24]:
crosstab

spending_flag  High spender  Low spender
region                                  
Midwest              155975      7441350
Northeast            108225      5614511
South                209691     10582194
West                 160354      8132559

In [11]:
# To find the total number of Instacart users per region

df_merge_PII.groupby('region').agg({'user_id':['count']})

user_id
              count
region             
Midwest     7597325
Northeast   5722736
South      10791885
West        8292913

# South region customers are the highest and lowest spenders 

 - Region make up about 38% of the country's population according to the US Census
 - South region also has the highest number of Instacart users. The number of users per region is directly proportional to the population.

 Source: https://www.census.gov/popclock/data_tables.php?component=growth

In [26]:
# Find percentage of high spenders to total number of users in Midwest region

print(155975 / 7597325)

0.020530252424373053


In [27]:
# Find percentage of high spenders to total number of users in Northeast region

print(108225 / 5722736)

0.018911408808653763


In [28]:
# Find percentage of high spenders to total number of users in South region

print(209691 / 10791885)

0.019430433144904714


In [29]:
# Find percentage of high spenders to total number of users in West region

print(160354 / 8292913)

0.019336269414619446


# Percent of high spenders versus total number of users per region

- The percent per region is very close to each other between 2-1.8 percent.

# 5. Exclude low activity customers

In [12]:
# Create flag for low activity customers

df_merge_PII.loc[df_merge_PII['max_order'] < 5, 'exclusion_flag'] = 'Low activity customer'

In [13]:
# Create flag for regular customers

df_merge_PII.loc[df_merge_PII['max_order'] >= 5, 'exclusion_flag'] = 'Regular customer'

In [32]:
# Check count under exclusion_flag 

df_merge_PII['exclusion_flag'].value_counts(dropna = False)

Regular customer         30964564
Low activity customer     1440295
Name: exclusion_flag, dtype: int64

In [14]:
# Exclude low activity customers then create new dataframe with only Regular Customer

df_final = df_merge_PII[df_merge_PII['exclusion_flag'] == 'Regular customer']

In [34]:
# Check number of rows

df_final.shape

(30964564, 32)

In [35]:
# Check columns in the df_final 

df_final.tail(5)

order_id user_id  order_number  order_dow  order_hour_of_day  \
32404850  3308056  106143            22          4                 20   
32404851  2988973  106143            23          2                 22   
32404852      930  106143            24          6                 12   
32404853   467253  106143            25          6                 16   
32404854   156685  106143            26          4                 23   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
32404850                  10.000      19675                  1          1   
32404851                   5.000      19675                  1          1   
32404852                   4.000      19675                  1          1   
32404853                   7.000      19675                  1          1   
32404854                   5.000      19675                  1          1   

                         product_name aisle_id department_id  prices  \
32404850  Organic Raspberry Black Tea       94             7  10.700   
32404851  Organic Raspberry Black Tea       94             7  10.700   
32404852  Organic Raspberry Black Tea       94             7  10.700   
32404853  Organic Raspberry Black Tea       94             7  10.700   
32404854  Organic Raspberry Black Tea       94             7  10.700   

            price_range_loc      activity busiest_period_of_day  max_order  \
32404850  Mid-range product  Slowest days        Average orders         26   
32404851  Mid-range product  Regular days        Average orders         26   
32404852  Mid-range product  Regular days           Most orders         26   
32404853  Mid-range product  Regular days           Most orders         26   
32404854  Mid-range product  Slowest days        Average orders         26   

              loyalty_flag  avg_order spending_flag  ord_med_freq  \
32404850  Regular customer     10.700  High spender         6.500   
32404851  Regular customer     10.700  High spender         6.500   
32404852  Regular customer     10.700  High spender         6.500   
32404853  Regular customer     10.700  High spender         6.500   
32404854  Regular customer     10.700  High spender         6.500   

            order_freq_flag gender   state  age date_joined  dependents  \
32404850  Frequent customer   Male  Hawaii   25   5/26/2017           0   
32404851  Frequent customer   Male  Hawaii   25   5/26/2017           0   
32404852  Frequent customer   Male  Hawaii   25   5/26/2017           0   
32404853  Frequent customer   Male  Hawaii   25   5/26/2017           0   
32404854  Frequent customer   Male  Hawaii   25   5/26/2017           0   

         fam_status  income _merge region    exclusion_flag  
32404850     single   53755   both   West  Regular customer  
32404851     single   53755   both   West  Regular customer  
32404852     single   53755   both   West  Regular customer  
32404853     single   53755   both   West  Regular customer  
32404854     single   53755   both   West  Regular customer

# 6. Export adjusted dataframe 

In [15]:
# export df_final to pkl 

df_final.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'Instacart_final.pkl'))